src/Microsoft.Azure.WebJobs.Host/Loggers/FunctionInstanceLogger.cs

```diff
@@ -17,56 +17,215 @@ internal class FunctionInstanceLogger : IFunctionInstanceLogger
-            Log(LogLevel.Information, message.Function, message.FunctionInstanceId, traceMessage);
-            if (message.TriggerDetails != null && message.TriggerDetails.Count != 0)
+            var logger = GetLogger(message.Function);
+            if (logger != null)
             {
-                LogTemplatizedTriggerDetails(message);
+                LogFunctionStarted(logger, message.Function.ShortName, message.FormatReason(), message.FunctionInstanceId, null);
+
+                if (message.TriggerDetails != null && message.TriggerDetails.Count != 0)
+                {
+                    LogTemplatizedTriggerDetails(logger, message);
+                }
             }
-            return Task.FromResult<string>(null);
+
+            return StringNullTask;
         }
 
-        private void LogTemplatizedTriggerDetails(FunctionStartedMessage message)
+        private void LogTemplatizedTriggerDetails(ILogger logger, FunctionStartedMessage message)
         {
-            var templateKeys = message.TriggerDetails.Select(entry => $"{entry.Key}: {{{entry.Key}}}");
-            string messageTemplate = "Trigger Details: " + string.Join(", ", templateKeys);
-            string[] templateValues = message.TriggerDetails.Values.ToArray();
-            Log(LogLevel.Information, message.Function, messageTemplate, templateValues);
+            var triggerDetails = message.TriggerDetails;
+            var key = message.Function.FullName;
+
+            if (_triggerLoggers.TryGetValue(key, out var log))
+            {
+                if (log(logger, triggerDetails))
+                {
+                    return;
+                }
+
+                var newLog = BuildTriggerLogger(triggerDetails);
+                newLog(logger, triggerDetails); // must log, it's build on this trigger
+
+                _triggerLoggers.TryUpdate(key, newLog, log); // try replace, may fail if another thread wrote sth
+            }
+            else
+            {
+                var newLog = BuildTriggerLogger(triggerDetails);
+                newLog(logger, triggerDetails); // must log, it's build on this trigger
+                _triggerLoggers.TryAdd(key, newLog);
+            }
         }
 
-        private void Log(LogLevel level, FunctionDescriptor descriptor, string message, params object[] args)
+        private static Func<ILogger, IDictionary<string, string>, bool> BuildTriggerLogger(IDictionary<string, string> triggerDetails)
         {
-            ILogger logger = _loggerFactory?.CreateLogger(LogCategories.CreateFunctionCategory(descriptor.LogName));
-            logger?.Log(level, 0, message, args);
+            var keys = triggerDetails.Keys.ToArray();
+
+            var templateKeys = keys.Select(key => $"{key}: {{{key}}}");
+            var messageTemplate = "Trigger Details: " + string.Join(", ", templateKeys);
```

Comment:
It won't throw, as there's a check for finding the right logger and the logger testing whether all the values are present. Speaking of fancy-ness 😜 , I was on the edge of leaving it with no structure. Then I thought that it might be worth it to structure it properly. I have no strong opinion in here but I like being structured in general. What do we do?

Reply:
Oh, I do see your checks now. I also would love to have structured logging wherever possible. However, since we don't log this as structured today, I'd opt to keep it simple at first. Then we can come back and iterate if we'd like to.